### PR TITLE
feat!: Update to acvm 0.12.0

### PR DESCRIPTION
**Pull request by TomAFrench**
Updates the ACVM dependency to current master of ACVM.

This has shown an unfortunate effect of https://github.com/noir-lang/acvm/pull/257 as we need to reconstruct the `BlackBoxFuncCall` which gets deconstructed in ACVM in order to call back into the solver functions.

### CODE REVIEW

File: src/acvm_interop/pwg.rs
Comment (phated) [inline src/acvm_interop/pwg.rs:35]: I think the ACVM functions need to be changed. It seems weird to be building this struct to call them.

Comment (TomAFrench) [reply src/acvm_interop/pwg.rs:35]: Yep, I made https://github.com/noir-lang/acvm/pull/268 to address this.

**Comment by phated** [comment]
I rebased this because I need it as a basis on my CRS changes that'll land in 0.12

**Comment by TomAFrench** [comment]
Yep, I'm happy for anyone to update this to target the ACVM master as needed.

**Comment by kevaundray** [comment]
This can now be updated to target 0.12 on crates.io

**Comment by phated** [comment]
A lot of folks have touched this, so I'm going to request reviews from folks that haven't (such as @kevaundray @kobyhallx and @vezenovm)

File: src/acvm_interop/proof_system.rs
Comment (vezenovm) [inline src/acvm_interop/proof_system.rs:26]: We don't support `Block`? cc'ing @guipublic

File: src/crs.rs
Comment (vezenovm) [inline src/crs.rs:31]: Should we add a sanity check that g1_and is greater than g1_start? Just in case `((num_points + 1) * 64) - 1` is smaller than `self.g1_data.len()`?

**Review by vezenovm — COMMENTED** [review]
A couple minor comments. Not really familiar with the nix stuff so I defer to @phated and @kobyhallx on that

File: src/acvm_interop/common_reference_string.rs
Comment (phated) [reply src/acvm_interop/common_reference_string.rs:5]:
```suggestion
// TODO(#185): Ensure CRS download works in JS
```
I'm unifying them and I tested it browser and it works; just need to upstream a bunch of needed changes from my WIP

File: src/crs.rs
Comment (phated) [reply src/crs.rs:31]: We actually do the num_points check in the composer API, but I'm fine wrapping this in an `if`. Pushing now.

Comment (vezenovm) [reply src/crs.rs:31]: Ah I missed that you do the `num_points` check there. Perhaps just note it in the method then? It seems unlikely that anyone else would ever be consuming the CRS API

File: src/acvm_interop/proof_system.rs
Comment (kevaundray) [reply src/acvm_interop/proof_system.rs:26]: It doesn't matter that much right now since the current Noir code will just use the block opcode anyways.

It will matter more depending on how 1282 is implemented. Block is a bit weird because its sort of like the fallback implementation for ROM/RAM

Comment (TomAFrench) [reply src/acvm_interop/proof_system.rs:26]: Sounds like this PR is good to go then?

Comment (kevaundray) [reply src/acvm_interop/proof_system.rs:26]: Yeah if that was the only blocker, then we can tenatively merge

**Review by kevaundray — APPROVED** [review]

**Pull request merged**
